src/search.py

```diff
@@ -0,0 +1,62 @@
+from collections import deque
+import st
+
+def LCP(a: str, b: str) -> int:
+    """
+    Finds the longest common prefix of a and b
+    
+    returns: 
+    If there is no prefix returns 0 otherwise returns length of prefix
+    """
+    i = 0
+    while(i < len(a) and i < len(b)):
+        if a[i] != b[i]:
+            return i
+        i+=1
+    return i
+
+def search(root: st.Node, pattern: str, string: str):
+    """
+    Looks for a pattern in a string using a suffix tree of string.
+    """
+    #Avoid copies while slicing with memoryviews
+
+    patternView = memoryview(pattern.encode())
+    stringView = memoryview(string.encode())
```

Comment:
Clever. I love it! However... `string.encode()` returns a new object, so I am pretty sure you copy the strings here.

```python
>>> x = 'foo'
>>> id(x)
4492696496
>>> id(x.encode())
4491140224
```

For the pattern, this is not a problem. We expect to spend $O(m)$ for a search anyway, so copying the pattern up front won't add to that. It is also quite often something we need anyway, as there are some benefits to mapping strings to smaller alphabets than the full Unicode, and then we need to map the pattern in the same way whenever someone wants to search.

For the string, that is more of an issue. It will cost $O(n)$, and for a search, we don't want to spend more time than $O(m)$ (and for the construction algorithm, we really must keep it shorter to get $O(n)$ in something like McCreight's algorithm).

The good news, though, is that you can just make the memory view when you build the tree and keep it there. Then you don't need to recreate it in searches, and, presto, you are back to $O(m)$. You then want two different search methods, one when the pattern is a `str` and one when it is a `memoryview`, but you can easily handle that in the same function

```python
def search(tree: WhateverYourTreeTypeIs, pattern: str | memoryview):
   root = tree.root
   stringView = tree.string
   patternView = pattern if instanceof(pattern, memoryview) else memoryview(pattern.encode())
   ...
```